File: solution53.java


public class solution53 {
    public int maxSubArray(int[] nums) {
        int res = Integer.MIN_VALUE;
        if (nums == null || nums.length < 1) {
            return res;
        }
        int[] dp = new int[nums.length];
        for (int i = 0; i < nums.length; i++) {
            for (int j = i; j < nums.length; j++) {
                if (i == j) {
                    dp[j] = nums[i];
                } else {
                    dp[j] = dp[j - 1] + nums[j];
                }
                if (dp[j] > res) {
                    res = dp[j];
                }
            }
        }
        return res;
    }
}

File: Cxx/solution53.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
   public:
    int maxSubArray(vector<int>& nums) {
        if (nums.size() < 1) {
            return 0;
        }
        int res = INT_MIN, dp = -1;
        for (int i = 0; i < nums.size(); i++) {
            if (dp > 0) {
                dp = dp + nums[i];
            } else {
                dp = nums[i];
            }
            if (dp > res) {
                res = dp;
            }
        }
        return res;
    }
};